File: backend/app/models/response/experiments.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel

from app.models.domain.enums import AlgorithmEnum, ExperimentStatusEnum


class ExperimentCreateResponse(BaseModel):
    experiment_id: str
    status: ExperimentStatusEnum
    message: str


class MetricsResult(BaseModel):
    accuracy: float | None = None
    precision: float | None = None
    recall: float | None = None
    f1_score: float | None = None


class ExperimentResult(BaseModel):
    metrics: MetricsResult
    confusion_matrix: list[list[int]]
    plots: dict[str, str]  # {"decision_boundary": "url", ...}


class ExperimentStatusResponse(BaseModel):
    id: str
    status: ExperimentStatusEnum
    algorithm: AlgorithmEnum | None = None
    hyperparameters: dict[str, Any] | None = None
    dataset_name: str | None = None
    dataset_id: str | None = None
    created_at: datetime | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    result: ExperimentResult | None = None
    error_message: str | None = None  # for failed status

File: frontend/src/types/experiments.ts
export interface CanvasPoint {
  x: number;
  y: number;
  class: string; // backend expects "class"
}

export interface CanvasDataset {
  type: 'canvas';
  name: string;
  points: CanvasPoint[];
  feature_names: string[];
}

export interface CreateExperimentRequest {
  dataset: CanvasDataset; // initially only canvas; later we'll support uploaded/builtin via union
  algorithm: string;
  hyperparameters: Record<string, unknown>;
  target_column: string;
}

export interface ExperimentCreateResponse {
  experiment_id: string;
  status: string;
  message: string;
}

export interface Metrics {
  accuracy: number;
  precision: number;
  recall: number;
  f1_score: number;
}

export interface ExperimentResult {
  metrics: Metrics;
  confusion_matrix: number[][];
  plots: {
    decision_boundary?: string;
    confusion_matrix_heatmap?: string;
  };
}

export interface ExperimentStatus {
  id: string;
  status: 'pending' | 'running' | 'completed' | 'failed';
  algorithm?: string;
  hyperparameters?: Record<string, unknown>;
  dataset_name?: string;
  dataset_id?: string;
  created_at?: string;
  started_at?: string;
  completed_at?: string;
  result?: ExperimentResult;
  error_message?: string;
}
